---
title: "SIB Days 2020 - Virtual Conference - June 8, 2020"
subtitle: "Dataset: Breast Cancer"
author:
  - Patrick Roelli, Computational Biologist 2 - Computational Biology^[10x Genomics, patrick.roelli@10xgenomics.com ]
  - Stefania Giacomello, Computational Biologist 2 - Computational Biology^[10x Genomics, stefania.giacomello@10xgenomics.com]
  - Stephen Williams, Senior Scientist - Computational Biology^[10x Genomics, stephen.williams@10xgenomics.com]
date: 'Last Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_notebook:
    code_folding: none
    theme: journal
    toc: yes
    toc_depth: 3
    toc_float: yes
---

[![10x Home](https://github.com/stephenwilliams22/SIB_2020_Workshop/raw/master/images/10x%20homepage.png)](https://www.10xgenomics.com/)

# **Introduction**

Welcome to the **SIB Days 2020 - virtual conference** Spatial Transcriptomics workshop by 10x genomics!

The purpose of this tutorial will be to walk users through some of the steps necessary to explore data produced by the 10x Genomics Visium Spatial Gene Expression Solution and the [Spaceranger pipeline](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger). We will investigate the data sets  are all freely available from [10x Genomics](https://support.10xgenomics.com/spatial-gene-expression/datasets).

[Seurat Tutorial](https://satijalab.org/seurat/v3.1/spatial_vignette.html)

**Things to know about this workshop**

1. All files that will be used can be found at: `/mnt/libs/shared_data/`

[![Terminal](https://github.com/stephenwilliams22/SIB_2020_Workshop/raw/master/images/Terminal.png)]

2. Getting started with R and Visium data outside of Seurat at: https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/rkit
3. Reference genome for all samples is GRCh38/mm10
4. All 10x software including [Spaceranger](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger), and [Loupe Browser](https://support.10xgenomics.com/spatial-gene-expression/software/visualization/latest/what-is-loupe-browser), can be downloaded from the [10x Support Site](https://support.10xgenomics.com/)
5. Most Seurat figures can be controlled with patchwork because they are fundamentally ggplot objects. Some really nice examples can be found on the [Patchwork GitHub](https://patchwork.data-imaginist.com/articles/guides/layout.html)


# **Exploring Visium Data with Seurat**

## Load our packages
```{r Libraries, echo=TRUE, message=FALSE, warning=FALSE}
library(Seurat)
library(ggplot2)
library(patchwork)
library(dplyr)
library(RColorBrewer)
```

## Loading data in a Seurat object

Real Dataset for the tutorial
```{r eval=FALSE}
breast_cancer <- Load10X_Spatial(data.dir = "/mnt/libs/shared_data/human_breast_cancer_1/outs/",
                filename = "V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5")
```


Same data just internal to 10x
```{r}
breast_cancer <- Load10X_Spatial(data.dir = "/mnt/analysis/marsoc/pipestances/HWHTFDSXX/SPATIAL_RNA_COUNTER_PD/163086/HEAD/outs/", slice = "slice1")
```

There are a bunch of data sets hosted by the Satija lab in the [Seurat Data Package](https://github.com/satijalab/seurat-data).

# **Results**
## QC
Let's have a look at some basic QC information. Keep in mind that most Seurat plots are ggplot object and can be manipulated as such.

Counts = UMI
Features = Genes
```{r, fig.width=10,fig.height=8, warning=FALSE}
plot1 <- VlnPlot(breast_cancer, features = "nCount_Spatial", pt.size = 0.1) + 
  ggtitle("UMI") +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend()

plot2 <- VlnPlot(breast_cancer, features = "nFeature_Spatial", pt.size = 0.1) + 
  ggtitle("Genes") +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend()

plot3 <- SpatialFeaturePlot(breast_cancer, features = "nCount_Spatial") + 
  theme(legend.position = "right")

plot4 <- SpatialFeaturePlot(breast_cancer, features = "nFeature_Spatial") +
  theme(legend.position = "right")

plot1 + plot2 + plot3  + plot4 + plot_layout(nrow = 2, ncol = 2)
```

## Normalization

Spaceranger does normalization for clustering and DE but does not return that normalized matrix

Pre-normalization 
Raw UMI counts
```{r, fig.width=10, warning=FALSE}
SpatialFeaturePlot(breast_cancer, features = c("ERBB2", "CD8A"))
```

### SE transform

+  This will take ~3-4 min. 

Don't worry about `reachediteration limit` warnings. See https://github.com/ChristophH/sctransform/issues/25 for discussion

Default assay will now be set to SCT
```{r, warning=FALSE}
breast_cancer <- SCTransform(breast_cancer, assay = "Spatial", verbose = TRUE)
```

```{r, fig.width=10, warning=FALSE}
SpatialFeaturePlot(breast_cancer, features = c("ERBB2", "CD8A"))
```


From Seurat: 

The default parameters in Seurat emphasize the visualization of molecular data. However, you can also adjust the size of the spots (and their transparency) to improve the visualization of the histology image, by changing the following parameters:

    pt.size.factor- This will scale the size of the spots. Default is 1.6
    alpha - minimum and maximum transparency. Default is c(1, 1).
    Try setting to alpha c(0.1, 1), to downweight the transparency of points with lower expression



```{r, fig.width=10, warning=FALSE}
p1 <- SpatialFeaturePlot(breast_cancer, features = "IGFBP5", pt.size.factor = 1)+ 
  theme(legend.position = "right") +
  ggtitle("Actual Spot Size")
p2 <- SpatialFeaturePlot(breast_cancer, features = "IGFBP5")+ 
  theme(legend.position = "right") +
  ggtitle("Scaled Spot Size")
p1 + p2
```

### Dimensionality reduction, clustering, and visualization

We can then proceed to run dimensionality reduction and clustering on the RNA expression data, using the same workflow as we use for scRNA-seq analysis.

Some of these processes can be parallized

```{r eval=FALSE, include=TRUE}
library(future)
# check the current active plan
plan()
```

```{r eval=FALSE, include=TRUE}
# change the current plan to access parallelization
plan("multiprocess", workers = 4)
plan()
```


The default UMAP calculation is performed with the [R-based UWOT](https://cran.r-project.org/web/packages/uwot/index.html) library However, you can run UMAP in python via reticulate library and `umap-learn`. We have found that for smaller data sets (<= 10k cells/spots) UWOT is great. For much larger data sets (100k + cells/spots) `umap-learn` can be a faster option.

```{r}
breast_cancer <- RunPCA(breast_cancer, assay = "SCT", verbose = FALSE)
breast_cancer <- FindNeighbors(breast_cancer, reduction = "pca", dims = 1:30)
breast_cancer <- FindClusters(breast_cancer, verbose = FALSE)
breast_cancer <- RunUMAP(breast_cancer, reduction = "pca", dims = 1:30)
```

Now let's have a look at the clustering

```{r, fig.width=8}
DimPlot(breast_cancer, reduction = "umap", label = FALSE,) +
  labs(color = "Cluster")
```

```{r, fig.width=10,fig.height=10, warning=FALSE}
p1 <- DimPlot(breast_cancer, reduction = "umap", label = TRUE) +
  labs(color = "Cluster")
p2 <- SpatialDimPlot(breast_cancer, label = TRUE, label.size = 3) +
  labs(fill = "Cluster")

p1 + p2 + plot_annotation(
  title = 'Clustering in UMAP and Tissue Space',
  caption = 'Processed by Spaceranger 1.1\nNormilization and Clustering by Seurat'
) + plot_layout(nrow = 1)
```

I don't really like these colors so let's change them
```{r}
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
```

```{r, fig.width=10, fig.height=10, warning=FALSE}
p1 <- DimPlot(breast_cancer, reduction = "umap", label = TRUE) +
  labs(color = "Cluster") + 
  scale_color_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green"))

p2 <- SpatialDimPlot(breast_cancer, label = TRUE, label.size = 3) +
  labs(fill = "Cluster")+ 
  scale_fill_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green"))

p1 + p2 + plot_annotation(
  title = 'Clustering in UMAP and Tissue Space',
  caption = 'Processed by Spaceranger 1.1\nNormilization and Clustering by Seurat'
) + plot_layout(nrow = 1)
```

If interested you can also now look at UMI and Gene counts per cluster as well
```{r}
plot1 <- VlnPlot(breast_cancer, features = "nCount_Spatial", pt.size = 0.1) + 
  ggtitle("UMI") +
  scale_fill_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green"))+
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend()

plot2 <- VlnPlot(breast_cancer, features = "nFeature_Spatial", pt.size = 0.1) + 
  ggtitle("Genes") +
  scale_fill_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green"))+
  theme(axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend()

plot1 + plot2
```


Now let's take a look at at a gene of interest with violin plots but also in image space.
```{r, fig.width=8, fig.height=8}
p1 <- VlnPlot(breast_cancer, features = "IGFBP5", pt.size = 0.1) + 
  ggtitle("IGFBP5") +
  scale_fill_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green"))+
  theme(axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend() +
  stat_summary(fun=mean, geom="point", shape=23, size=4, color="red")

p2 <- SpatialFeaturePlot(breast_cancer, features = "IGFBP5")+ 
  theme(legend.position = "right")

p3 <- SpatialDimPlot(breast_cancer, label = TRUE, label.size = 3) +
  labs(fill = "Cluster")+ 
  scale_fill_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green")) +
  NoLegend()

row1 <- p2 + p3 + plot_layout(nrow = 1)
row1 + p1+ plot_layout(nrow = 2, widths = c(0.5, 0.5))
```

We can also look at these data interactively
```{r eval=FALSE, include=TRUE}
LinkedDimPlot(breast_cancer)
```

## Spatially variable features

First we'll identify deferentially expressed genes. 

Parallelization helps here too let's make sure our plan is still intact

```{r, eval=FALE, include=TRUE}
plan()
```
`- call: plan("multiprocess", workers = 4)` indicates that it is

Looks like we have some very DE genes for clusters 4 and 11


clarify what ident.1 = 4, ident.2 = 6 are for
```{r, fig.width=10, warning=FALSE, message=FALSE}
de_markers <- FindMarkers(breast_cancer, ident.1 = 4, ident.2 = 6)
de_markers
```

## Identify the most up-regulated and down-regulated genes
```{r}
de_markers_up <- de_markers %>%
  tibble::rownames_to_column(var = "Gene") %>% 
  arrange(-avg_logFC)

de_markers_down <- de_markers %>%
  tibble::rownames_to_column(var = "Gene") %>% 
  arrange(avg_logFC)
```


### Most up-regulated genes
```{r, fig.width=10, fig.height=10, warning=FALSE}
SpatialFeaturePlot(object = breast_cancer, features = de_markers_up$Gene[1:10], alpha = c(0.1, 1), ncol = 3)
```

### Most down-regulated genes
```{r, fig.width=10, fig.height=10, warning=FALSE}
SpatialFeaturePlot(object = breast_cancer, features = de_markers_down$Gene[1:10], alpha = c(0.1, 1), ncol = 3)
```

what are the top variable features?
```{r}
VariableFeatures(breast_cancer)[1:10]
```

what are the top DE genes?
```{r}
rownames(de_markers)[1:10]
```

## Spatially Variable Genes

So what about spatial enrichment? 

Some methods

1. [Trendsceek](https://www.nature.com/articles/nmeth.4634)
2. [Splotch](https://www.biorxiv.org/content/10.1101/757096v1)
3. [SPARK](https://www.nature.com/articles/s41592-019-0701-7)
4. [SpatialDE](https://www.nature.com/articles/nmeth.4636)
  + We have found this implementation not to be very effective. It's also not under active development


Using the top 100 variable genes find spatially enriched ones. Note that in the Seurat Spatial Tutorial they use 1000 genes. You can also use all genes but that will take a long time. Using a calculation of [Morans I](https://en.wikipedia.org/wiki/Moran%27s_I) can sometimes be a faster approach, especially if you are using parallization.


```{r}
breast_cancer <- FindSpatiallyVariableFeatures(breast_cancer, 
                                               assay = "SCT", 
                                               slot = "scale.data", 
                                               features = VariableFeatures(breast_cancer)[1:100],
                                               selection.method = "markvariogram", verbose = TRUE)
```

Have a look at the spatially variable genes calculated by `markvariogram` ordered from most variable to least variable
```{r}
SpatiallyVariableFeatures(breast_cancer, selection.method = "markvariogram", decreasing = TRUE)
```

```{r, fig.width=10, fig.height=8, warning=FALSE}
top.features_trendseq <- head(SpatiallyVariableFeatures(breast_cancer, selection.method = "markvariogram"), 8)
SpatialFeaturePlot(breast_cancer, features = top.features_trendseq, ncol = 4, alpha = c(0.1, 1))
```

Moran's I implementation. For other spatial data types the x.cuts and y.cuts determines the grid that is laid over the tissue in the capture area. Here we'll remove those
```{r}
breast_cancer <- FindSpatiallyVariableFeatures(breast_cancer, 
                                               assay = "SCT", 
                                               slot = "scale.data", 
                                               features = VariableFeatures(breast_cancer)[1:100],
                                               selection.method = "moransi")
```


Have a look at the spatially variable genes calculated by `moransi` ordered from most variable to least variable

```{r}
SpatiallyVariableFeatures(breast_cancer, selection.method = "moransi", decreasing = TRUE)
```

```{r, fig.width=10, fig.height=8, warning=FALSE}
top.features_moransi <- head(SpatiallyVariableFeatures(breast_cancer, selection.method = "moransi"), 8)
SpatialFeaturePlot(breast_cancer, features = top.features_moransi, ncol = 4, alpha = c(0.1, 1))
```

We can see that the results are slightly different. So let's take a look at why.

```{r}
spatially_variable_genes <- breast_cancer@assays$SCT@meta.features %>%
  tidyr::drop_na()

spatially_variable_genes
```
You can see the two methods show 
```{r}
mm_cor <- cor.test(spatially_variable_genes$moransi.spatially.variable.rank, spatially_variable_genes$markvariogram.spatially.variable.rank)
ggplot(spatially_variable_genes, aes(x=moransi.spatially.variable.rank,y=markvariogram.spatially.variable.rank))+
  geom_point()+
  geom_smooth()+
  xlab("Morans I Rank")+
  ylab("Markvariogram Rank")+
  annotate("text", x = 25, y = 75, label = paste("Pearson's Correlation\n", round(mm_cor$estimate[1], digits = 2), sep = ""))+
  theme_bw()
```

## Cancer annotations

Where are these genes being expressed relative to pathologist annotation?

```{r, fig.height=8, fig.width=8}
ca <- readbitmap::read.bitmap("images/Breast Cancer Path.png")
# in the tutorial
# ca <- readbitmap::read.bitmap('/mnt/libs/shared_data/human_breast_cancer_1/images/Breast_Cancer_Path.png')
plot(0:1,0:1,type="n",ann=FALSE,axes=FALSE)
rasterImage(ca,0,0,1,1)
```

Looks like the  Matrix Gla protein ( _MGP_ ) gene is enriched in Ductal Carcinoma _In Situ_. Not a lot is known about _MGP_ in the context of cancer but it looks like it could be an interesting novel gene to investigate with regard to Ductal Carcinoma _In Situ_.

```{r, warning=FALSE}
SpatialFeaturePlot(object = breast_cancer, features = "MGP", alpha = c(0.1, 1), ncol = 3)
```

# Multiple Sections

for workshop
```{r eval=FALSE}
breast_cancer_2 <- Load10X_Spatial(data.dir = "/mnt/libs/shared_data/human_breast_cancer_2/outs/",
                filename = "V1_Breast_Cancer_Block_A_Section_2_filtered_feature_bc_matrix.h5")
```
for 10x
```{r}
breast_cancer_2 <- Load10X_Spatial(data.dir = "/mnt/analysis/marsoc/pipestances/HWHTFDSXX/SPATIAL_RNA_COUNTER_PD/163087/HEAD/outs/", 
                                   slice = "slice2")
```

```{r}
breast_cancer_merged <- merge(breast_cancer, breast_cancer_2)
```

Counts = UMI
Features = Genes
```{r, fig.width=10,fig.height=10, warning=FALSE}
plot1 <- SpatialFeaturePlot(breast_cancer_merged, features = "nCount_Spatial") + 
  theme(legend.position = "right")

plot2 <- SpatialFeaturePlot(breast_cancer_merged, features = "nFeature_Spatial") +
  theme(legend.position = "right")

plot1 + plot2 + plot_layout(nrow = 2)
```

## Normalize merged data

This will take ~3min

Default assay will now be set to SCT
```{r, warning=FALSE}
breast_cancer_merged <- SCTransform(breast_cancer_merged, assay = "Spatial", verbose = TRUE)
```

```{r, fig.width=10, warning=FALSE}
SpatialFeaturePlot(breast_cancer_merged, features = c("ERBB2", "CD8A"))
```

# Single cell/nuclei ingegration

Here we have a preprocessed Seurat object with 10k nuclei annotated from a breast cancer sample. Don't bother too much with the details of how this data was generated they don't particularly matter for our purposes. 

## Load the Seurat object
10x infrastructure
```{r}
bc_snRNA <- readRDS("/mnt/home/stephen.williams/yard/Odin/SIB_2020_Workshop/bc_snRNA.rds")
```

For the workshop
```{r eval=FALSE, include=TRUE}
bc_snRNA <- readRDS("/mnt/libs/shared_data/bc_snRNA.rds")
```


```{r}
bc_snRNA
```

```{r, fig.width=10}
DimPlot(bc_snRNA, group.by = "subclass", label = FALSE)
```
